ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,341 @@
+                  |> Int64.of_string
+                  |> Mtime.of_uint64_ns
+                in
+                Lwt.return (Some (Content ((uuid, timestamp), latest)))
+          )
+          tpms
+        |> Lwt_list.map_p Fun.id
+      in
+      Lwt.return (List.filter_map Fun.id actions)
+    in
+    let* contents = get_all_contents () in
+    let* () = resolve_actions push contents in
+    let@ () = with_lock queue.lock in
+    let* contents = get_all_contents () in
+    let* () =
+      match contents with
+      | [] ->
+          queue.state <- Direct ;
+          Logs_lwt.info (fun m -> m "%s: Cache clean; Going direct" __FUN)
+      | _ ->
+          Lwt.return_unit
+    in
+    Lwt.return_unit
+
+  let resolve queue push root = function
+    | Actions actions -> (
+        let* () = resolve_actions push actions in
+        let@ () = with_lock queue.lock in
+        match queue.state with
+        | Direct | Disengaged ->
+            Lwt.return_unit
+        | Engaged ->
+            let () =
+              if Lwt_bounded_stream.size queue.queue = 0 then
+                queue.state <- Direct
+            in
+            Lwt.return_unit
+      )
+    | Update_all ->
+        update_all queue push root
+    | Wait ->
+        (* Do not busy loop when the system can cope with the requests *)
+        Lwt_unix.sleep 0.2
+
+  let watch ~direct:push typ queue =
+    let root = cache_of typ in
+    let __FUN = __FUNCTION__ in
+    let rec loop () =
+      (* When the pushing side is disengaged it doesn't push events to the
+         queue, this means that trying to drain it completely would leave the
+         pulling side locked waiting when the queue is empty.
+           - Read the number of elements in the queue while draining it and
+             then switch to read the contents from the cache; or
+           - Switch immediately to reading the contents from cache and ignore
+             the contents of the queue by calling an specialized method in the
+             queue module to drain it.
+      *)
+      let get_action () =
+        let path (uuid, timestamp) =
+          root
+          // Uuidm.to_string uuid
+          // Mtime.(to_uint64_ns timestamp |> Int64.to_string)
+        in
+        let@ () = with_lock queue.lock in
+        match queue.state with
+        | Disengaged ->
+            let* () = Lwt.pause () in
+            Lwt.return Update_all
+        | Direct ->
+            let* () = Lwt.pause () in
+            Lwt.return Wait
+        | Engaged -> (
+            let* elem = Lwt_bounded_stream.get queue.queue in
+            match elem with
+            | None ->
+                raise (Failure "Other side closed channel, cannot continue")
+            | Some elem ->
+                Lwt.return (Actions [Content (elem, path elem)])
+          )
+      in
+      let* action = get_action () in
+      let* () = resolve queue push root action in
+      loop ()
+    in
+    loop
+end
+
+let setup typ direct =
```

Comment:
What happens when you restart xapi-guard, or the host? Will it try to find caches that haven't been sent to xapi yet and try to send them?

Reply:
Yes, this is why the cache is started in the "Disengaged mode" where the cache is flushed without reading the queue, while new contents can be pushed to it